src/wp-includes/class-wp-query.php

```diff
@@ -4755,6 +4722,53 @@ public function generate_postdata( $post ) {
 
 		return $elements;
 	}
+
+	/**
+	 * Generate cache key.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @global wpdb $wpdb WordPress database abstraction object.
+	 *
+	 * @param array  $args Query arguments.
+	 * @param string $sql  SQL statement.
+	 *
+	 * @return string Cache key.
+	 */
+	protected function generate_cache_key( array $args, $sql ) {
+		global $wpdb;
+
+		unset(
+			$args['cache_results'],
+			$args['fields'],
+			$args['lazy_load_term_meta'],
+			$args['update_post_meta_cache'],
+			$args['update_post_term_cache'],
+			$args['update_menu_item_cache'],
+			$args['suppress_filters']
+		);
+
+		array_walk_recursive(
+			$args,
+			function ( &$value ) use ( $wpdb ) {
+				$placeholder = $wpdb->placeholder_escape();
+				if ( is_string( $value ) && str_contains( $value, $placeholder ) ) {
+					$value = $wpdb->remove_placeholder_escape( $value );
+				}
+			}
```

Comment:
> are you good with the closure here?

Yes, think a closure as the callback to `array_walk_recursive()` is good/desirable. 

The only place where closures cannot be used is as callbacks to WP hooks (actions and filters) as they break the API. In all other cases, and especially as callbacks for built-in PHP functions, they seem the preferred way. Another exception, of course, would be when the callback function is more complex and needs to be tested. 

> it was less clear why the value was passed by reference.

Yes, passing by reference may look a bit unexpected. It is required as [`array_walk_recursive()`](https://www.php.net/manual/en/function.array-walk-recursive.php) does not return a changed array, it can change the original array. Perhaps an inline comment there would be good.

Reply:
I've been wondering if `$placeholder = $wpdb->placeholder_escape();` can be moved outside of the callback, before array_walk_recursive(). Currently it will be called fro each non-array element of the `$args` array. Don't think that's going to be a lot but..?

Reply:
I think so @azaozz, though no need to pass the `$placeholder` as a reference since it isn't changed within the closure.

Reply:
@azaozz @hellofromtonya Done in b354a5f6f1cd0e4a637c2282dbc710e167f5aca9

Reply:
> no need to pass the $placeholder as a reference since it isn't changed within the closure.

Yes, it is not changed inside the closure, but imho still need to be passed by reference. The thing is that the `use` keyword/language construct is actually three different things: 
1. to import in namespaces
2. to insert traits in classes
3. to inherit variables in closures
(I know, it's quite confusing).

In the third case the value of a variable that's defined in the parent scope and then imported in the closure is set when the code is parsed. This means that if the closure is inside a function, that variable will be imported when the file is loaded, not later on when the function is executed. 

So if the value of the var is set from a function call (like in this case) and it can change, the value of the imported var will not change. I've ran into problems with this before. The way to fix this is to pass the variable by reference. Then if it changes in the parent scope, it will also change in the closure. See: https://www.php.net/manual/en/functions.anonymous.php#example-184. Also note that `$wpdb` is a class instance that's always passed by reference so there's no problems with it.